lms/djangoapps/courseware/user_state_client.py

```diff
@@ -149,29 +162,45 @@ def get_many(self, username, block_keys, scope=Scope.user_state, fields=None):
                 continue
 
             state = json.loads(module.state)
-            state_length += len(module.state)
+            state_length = len(module.state)
 
-            self._ddog_histogram(evt_time, 'get_many.block_size', len(module.state))
+            # NOTE: should this line be before the check for empty state?
+            self._ddog_histogram(evt_time, 'get_many.block_size', state_length)
 
             # If the state is the empty dict, then it has been deleted, and so
             # conformant UserStateClients should treat it as if it doesn't exist.
             if state == {}:
                 continue
 
+            # collect statistics for DataDog/New Relic reporting
+            block_stats_by_type.setdefault(usage_key.block_type, {})
+            block_stats_by_type[usage_key.block_type].setdefault('count', 0)
+            block_stats_by_type[usage_key.block_type].setdefault('size', 0)
+            block_stats_by_type[usage_key.block_type]['count'] += 1
+            block_stats_by_type[usage_key.block_type]['size'] += state_length
+
+            # filter state on fields
             if fields is not None:
                 state = {
                     field: state[field]
                     for field in fields
                     if field in state
                 }
-            block_count += 1
             yield XBlockUserState(username, usage_key, state, module.modified, scope)
 
-        # The rest of this method exists only to submit DataDog events.
-        # Remove it once we're no longer interested in the data.
+        # The rest of this method exists only to submit DataDog and New Relic metrics.
         finish_time = time()
-        self._ddog_histogram(evt_time, 'get_many.blks_out', block_count)
-        self._ddog_histogram(evt_time, 'get_many.response_time', (finish_time - evt_time) * 1000)
+        duration = (finish_time - evt_time) * 1000  # milliseconds
+        total_block_count = sum([block_info['count'] for block_info in block_stats_by_type.values()])
+        total_state_length = sum([block_info['size'] for block_info in block_stats_by_type.values()])
+        self._ddog_histogram(evt_time, 'get_many.blks_out', total_block_count)
+        self._ddog_histogram(evt_time, 'get_many.response_time', duration)
+        self._nr_capture_metric('get_many', 'num_items', total_block_count)
```

Comment:
Ah, makes sense.  I assume the difference with the other example [0] you provided is that those functions only get called once per request/transaction by design (e.g. views), whereas the user state client does not have the same access pattern?

[0] https://github.com/edx/edx-platform/blob/dda92d13198b935f8ebaaaf1ab29b3f05c2ec6ea/common/lib/xmodule/xmodule/seq_module.py#L303